string_utils/fms_string_utils_binding.c

```diff
@@ -162,3 +162,20 @@ char* fms_find_my_string_binding(char** arr, int *n, char *find_me, int *np)
   strcpy(string_p, string);
   return string_p;
 }
+
+// Finds the number of unique strings in an array
+int fms_find_unique(char** arr, int *n)
+{
+  int i;
+  int nfind;
+
+  nfind=1;
+  //printf("n is %i", *n);
+  for(i=1; i<*n; i++){
+    //printf("Comparing %s and %s \n",arr[i], arr[i-1]);
+    if (strcmp(arr[i], arr[i-1]) != 0){ nfind = nfind + 1;}
+  }
+
+  //printf("nfind=%i",nfind);
+  return nfind;
+}
```

Comment:
Unless I'm misreading the logic, this only works for a sorted array.   To do this without a pre-sort, you'd need a nested loop structure.  You can test this by moving the fms_find_unique in your unit test to be line 59 prior to the sort.  If you want to stick with a single loop, you'd need add an fms_sort_this prior to the search-loop.

Reply:
This was only intended to work with a sorted array:
https://github.com/NOAA-GFDL/FMS/blob/3db45b1d78d4787fa47e52d8f121a4fdd4621d68/string_utils/fms_string_utils.F90#L71-L81

But the documentation should be more clear.

Reply:
You could add a call to the fms_sort_this function to ensure a sort is done prior to the comparison search.  That would remove the requirement to pass in only sorted arrays.

Reply:
I added the call

Reply:
Is there a quick way to test if it's sorted?  Is the sorting of a sorted array fast enough that it doesn't matter that you sort it again?